integration-tests/src/test/java/com/datastax/oss/driver/core/TCPFlowControlIT.java

```diff
@@ -0,0 +1,251 @@
+      List<Node> nonPausedNodes =
+          session.getMetadata().getNodes().values().stream()
+              .filter(n -> !pausedHostAddress.equals(n.getBroadcastRpcAddress().get()))
+              .collect(Collectors.toList());
+      assertThat(nonPausedNodes.size()).isEqualTo(2);
+
+      SIMULACRON_RULE.cluster().node(2).pauseRead();
+
+      List<CompletionStage<AsyncResultSet>> pendingRequests = new ArrayList<>();
+      for (int i = 0; i < NUMBER_OF_SUBMITTED_REQUESTS; i++) {
+        pendingRequests.add(session.executeAsync(SimpleStatement.newInstance(queryString)));
+      }
+
+      // Non-paused nodes should process the requests correctly
+      for (Node n : nonPausedNodes) {
+        Awaitility.await()
+            .until(
+                () -> {
+                  int inFlightRequests = getInFlightRequests(session, n);
+                  return inFlightRequests == 0;
+                });
+      }
+
+      // There are still requests that haven't been written
+      waitForWriteQueueToStabilize(session);
+      assertThat(getWriteQueueSize(session)).isGreaterThan(1);
+
+      SIMULACRON_RULE.cluster().resumeRead();
+
+      int numberOfFinished = 0;
+      for (CompletionStage<AsyncResultSet> result : pendingRequests) {
+        result.toCompletableFuture().get(1, TimeUnit.SECONDS);
+        numberOfFinished = numberOfFinished + 1;
+      }
+
+      int writeQueueSize = getWriteQueueSize(session);
+      assertThat(writeQueueSize).isEqualTo(0);
+      assertThat(numberOfFinished).isEqualTo(NUMBER_OF_SUBMITTED_REQUESTS);
+    }
+  }
+
+  @Test
+  public void should_timeout_requests_when_the_server_paused_reading_without_resuming()
+      throws InterruptedException {
+    byte[] buffer = new byte[10240];
+    Arrays.fill(buffer, (byte) 1);
+    ByteBuffer buffer10Kb = ByteBuffer.wrap(buffer);
+
+    String queryString =
+        String.format("INSERT INTO table1 (id) VALUES (%s)", ByteUtils.toHexString(buffer10Kb));
+    Query query = new Query(queryString, emptyList(), emptyMap(), emptyMap());
+
+    SIMULACRON_RULE.cluster().prime(when(query).then(noRows()));
+
+    try (CqlSession session = SessionUtils.newSession(SIMULACRON_RULE)) {
+      SIMULACRON_RULE.cluster().pauseRead();
+
+      List<CompletionStage<AsyncResultSet>> pendingRequests = new ArrayList<>();
+      for (int i = 0; i < NUMBER_OF_SUBMITTED_REQUESTS; i++) {
+        pendingRequests.add(session.executeAsync(SimpleStatement.newInstance(queryString)));
+      }
+
+      // Assert that there are still requests that haven't been written
+      waitForWriteQueueToStabilize(session);
+      assertThat(getWriteQueueSize(session)).isGreaterThan(1);
+
+      // do not resumeRead
+      for (CompletionStage<AsyncResultSet> result : pendingRequests) {
+        assertThatThrownBy(() -> result.toCompletableFuture().get())
+            .isInstanceOf(ExecutionException.class)
+            .hasCauseInstanceOf(DriverTimeoutException.class);
```

Comment:
Using the same driver class that I mentioned earlier:
```java
        assertThatStage(result)
            .isFailed(e -> assertThat(e).isInstanceOf(DriverTimeoutException.class));
```
Or you could do it in pure AssertJ:
```java
        assertThat(result).hasFailedWithThrowableThat().isInstanceOf(DriverTimeoutException.class);
```

Reply:
None of the above versions worked for me, the test hangs. As I don't have much time to investigate I will just leave the code as is now.

Reply:
Actually this test hangs very often. An assertion fails, then there is a problem with the session closing sequence that never ends.